feat: Allow type-definition of CustomError.data

nidomiro opened the pull request:
Allows typing of `CustomError.data` to store known additional Information in an predefined format.

The only caveat I can think of is with hydration.
There the type can be misleading due to Serialisation/Deserilisation with classes. But actual runtime-behaviour is the same as the current implementation.

goloroden commented:
@nidomiro Thanks for your PR, we'll review it and get back with feedback to you 😊

Just one quick question in the beginning: You changed the `.gitignore` file to exclude your IDE files. We prefer to have these settings in a global file, not a per-project file. Could you maybe undo these changes?

goloroden commented on .gitignore:
```diff
@@ -8,3 +8,6 @@
 # OS generated files and folders
 .DS_Store
 npm-debug.log
+
+# IDEs
+.idea
```
See comment above 😉

nidomiro replied:
it's changed in the new commit

yeldiRium commented:
I like the idea. However I have a suggestion to improve the usability a bit. With the current implementation I would create a custom error class like this:

```typescript
class CustomError extends defekt<'CustomError', { foo: string }>({ code: 'CustomError' }) {}
```

This is even more redundant than it already is, since now we have to pass the `TErrorName` type parameter to be able to specify the type parameter for the data property.
To improve this a bit I think we could switch the two type parameters:

```typescript
const defekt = function <TDataType = any, TErrorCode extends string>({
```

Which would result in this usage:

```typescript
class CustomError extends defekt<{ foo: string }>({ code: 'CustomError' }) {}
```

A drawback in my opinion is that this order of type parameters is not very intuitive, since the more important type parameter is at the back. But when actually using defekt, one almost never specifies the `TErrorCode` by passing it as a type parameter, so pushing it to the back is fine, I think.

@nidomiro what do you think about this suggestion? Also it would be great to have some tests for the new functionality. Testing purely type-system based changes it always a bit awkward, but a small test in which a newly typesafe property on an error's data is accessed would be enough, i think.

nidomiro commented:
@yeldiRium, I also thought about your suggestion. It would technically Introduce a breaking change, thats why I omitted the thought. In reality it shouldn't be a problem.
For me it's ok, that the more important parameter is not at the start, because it does simplify using the code.

Since `function <TDataType = any, TErrorCode extends string>` is not possible but `function <TDataType = any, TErrorCode extends string = ''>` is and produces the same types as before there are two ways of solving this:

1. Change the generic parameters of `CustomErrorConstructor`, `CustomError` and `defekt` to this new 'signature'
2. Change only the generic parameters of `defect` and keep the rest the same

I would say option one is more consistent and therefore the better one.

nidomiro commented:
@yeldiRium I've changed the code according to option 1 and now I'm searching for a way to make data a required attribute if chosen to do so.

nidomiro commented:
I currently cannot find a solution to this problem without breaking existing code. Therefore I won't continue searching for a solution in this PR.
If you have any idea on how to achieve this, I would love to hear it.

yeldiRium commented:
Hi @nidomiro,

thank you for your changes and for including a test case!

Also sorry for the late response. I tried a few ways to get the type-safety to work, but as you predicted when opening the PR, hydration is a problem. It is possible to enforce the `data` parameter on `ErrorConstructor` if a type parameter for it is set, but this unfortunately breaks the hydration logic and I was not able to get it to work.

I'm a bit annoyed by this, since with the current implementation devs still need to check whether an error has data, before accessing it. But at least that is consistent and type-safe. Except for hydration, since we cannot guarantee the content of the data property there.

But I still like the improvement over how it was before, so I'll be happy to merge it.

But before I do so, would you like to add yourself to the contributors in the `package.json`?

goloroden commented on lib/CustomError.ts:
```diff
@@ -1,13 +1,13 @@
-class CustomError<TErrorName extends string = string> extends Error {
-  public code: TErrorName;
+class CustomError<TDataType = any, TErrorCode extends string = string> extends Error {
```
One super-small thing: Can you maybe change `TDataType` to `TData` (and check that it's being consistent everywhere)?

nidomiro replied:
done ✅

goloroden reviewed (changes requested):
Hey 👋 

I know I'm late to the party, and I've only got a very small thing 😉

nidomiro commented:
Yea, hydration mostly just works for plain objects, but with classes (and therefore containing methods, prototypes, ...) or symbols it has to involve some user-action. I don't think this task can be automated in ts right now.
In another project we used neverthrow (for the Result functionality) instead of defekt. There I had written a custom hydrate-function, which accepted functions to hydrate the value, and the error. It was still a non-optimal solution, but it worked well enough for this use-case.

Maybe we find a valid solution - if something comes to my mid I'll create another pr.


I added myself to the contributors and changed TDataType to TData. From my point of view the pr is ready.
